### PR TITLE
Parse Account Summary section of pdf issue #98

### PR DESCRIPTION
**Pull request by jamesschlader**
This offers an opinionated method for parsing the Account Summary section of a Docket pdf. The parseAccountSummary function returns an object that summarizes all of the accounts with associated values for this Docket. All the comments and rubbish are removed during parsing.

Rudimentary tests are also included.

### CODE REVIEW

**Comment by joeldenning** [comment]
Here's a rough draft of an implementation that I think would be a lot easier for people to read and understand the code:

```js
const subsections = []

walkSection(lines, sections, "ACCOUNT SUMMARY", (line, index) => {
  const words = line.split(/\s/)
  const isSubsection = words.length > 0 && words[0].toUpperCase() === words[0];
  if (isSubsection) {
    subsections.push({name: line, lineNumber: index})
  }
})

const balances = subsections.map((subsection, arrayIndex) => {
  const endIndex = subsections[arrayIndex + 1].lineNumber // to-do make this work for last item in array, too

  let amountDue, amountPaid, costType

  for (let i = subsection.lineNumber; i < endIndex; i++) {
    const line = lines[i];
    const words = line.split(/\s/);

    if (line.includes('Amount Due')) {
      amountDue = words[words.length - 1];
    } else if (line.includes('Amount Paid') || line.includes('Paid In')) {
      amountPaid = words[words.length - 1];
    }

    costType = getCostType(line) || costType;
  }

  return {
    name: subsection.name,
    amountDue,
    amountPaid,
    costType,
  }
})

return balances

function getCostType(line) {
  if (line.includes('Restitution')) {
    return 'restitution'
  }
}
```

**Comment by jamesschlader** [comment]
New commits based on feedback.

File: backend/apis/parse-docket-pdf.utils.js [hunk 1]
```diff
@@ -89,13 +89,80 @@ function parseCharges(lines, sections) {
...
+          costType,
+          collection
+        };
+        return amountDue && returnObj;
+      })
+      .filter(result => {
+        return result;
+      });
+  }
+  function getCostType(line) {
```
Comment (joeldenning) [inline backend/apis/parse-docket-pdf.utils.js]: Can we make this function better to include detection of fines and fees (in addition to restitution)? For example, this line of code is for a fine:

https://github.com/JustUtahCoders/utahexpungements.org/blob/c9522469b27675264611ce65654147627a1566bd/backend/apis/docket-pdf-test-files/account-summary.js#L53

Fines, fees, and restitution are treated differently when determining eligibility, which is why it's important to try to determine what kind of charge it is, if possible.

File: backend/apis/docket-pdf-test-files/get-text.js [hunk 2]
```diff
@@ -0,0 +1,26 @@
+const fs = require("fs");
```
Comment (joeldenning) [inline backend/apis/docket-pdf-test-files/get-text.js]: Now that we have the UI that veronica created, is this still needed? If it is still useful to you or others, then let's keep it. But if Veronica's UI is what is being used to do this, then let's delete this file.

File: backend/apis/normal-docket-pdf.test.js [hunk 3]
```diff
@@ -29,3 +31,52 @@ describe("normal docket pdf file", () => {
     expect(actual1.dispositionDate).toBe("Month 00, 0000");
   });
 });
+
+describe("correctly and completely parses ACCOUNT SUMMARY section", () => {
```
Comment (joeldenning) [inline backend/apis/normal-docket-pdf.test.js]: Let's create two test suites (`describe()` calls) - one for each test file. The variable names inside of this test suite (`parsedSection`, for example) make it hard to understand which test file is being tested with each assertion.

Good job writing these tests, though. They are validating that the correct things are being extracted 👍

File: backend/apis/docket-pdf-test-files/account-summary.js [hunk 4]
```diff
@@ -0,0 +1,879 @@
...
+05-23-14 Case filed                                                             
+05-23-14 Judge MICHAEL W KWAN assigned.                                         
+05-23-14 ARRAIGNMENT scheduled on July 17, 2014 at 08:05 AM in MAIN 
+         COURTROOM with Judge KWAN.                                             
+05-23-14 **** PRIVATE **** Filed: Citation                              
+05-23-14 Note: information and summons to constable for service                 
+06-05-14 Note: Def calls to ask if there is a case against her as she 
+               has received solicitation letters from defense attorneys.        
+06-16-14 Filed return: SUMMONS                                                  
+                   Party Served: MIKE EVANS - FATHER
```
[the remark this answers is not in the document]
Comment (jamesschlader) [reply backend/apis/docket-pdf-test-files/account-summary.js]: Missed it. Will change.

File: backend/apis/parse-docket-pdf.utils.js [hunk 5]
```diff
@@ -89,13 +89,80 @@ function parseCharges(lines, sections) {
...
+    return section.name === "PROCEEDINGS" && section.lineNumber;
+  })[0].lineNumber;
+
+  function addAccounts(line, index) {
+    const words = line.split(/\s/);
+    const isSubsection =
+      words.length > 0 &&
+      words[0].toUpperCase() === words[0] &&
+      !words.includes("CASE") && // prevents CASE NUMBER lines from defining a subsection
+      words[1]; // prevents line new page diving line from defining a subsection;
```
[the remark this answers is not in the document]
Comment (jamesschlader) [reply backend/apis/parse-docket-pdf.utils.js]: The problem was not about new line characters or spaces. What was happening was that the new page line separator:
_____________________________________________________________________________________________________

was passing the easy case and being assigned as the value of the "name" property. That is, (1) there was at least one word (words.length > 0) because words[0] === "____...", (2) words[0].toUpperCase() === words[0] ("_".toUpperCase() returns "_"), (3) words[0] !== "CASE". (I added this condition to prevent the CASE NUMBER line from starting a new subSection.) Adding that last condition (words[1]) returned the right answer. 
"_".trim() returns "_".

I noticed after that the comment itself is very unclear. So sorry about the confusion.

File: backend/apis/parse-docket-pdf.utils.js [hunk 6]
```diff
@@ -89,13 +89,80 @@ function parseCharges(lines, sections) {
...
+            if (line.includes("Original")) {
+              originalAmountDue = words[words.length - 1];
+            } else if (line.includes("Amended")) {
+              amendedAmountDue = words[words.length - 1];
+            }
+            amountDue = words[words.length - 1];
+          } else if (line.includes("Amount Paid") || line.includes("Paid In")) {
+            amountPaid = words[words.length - 1];
+          }
+          // console.log(line);
```
[the remark this answers is not in the document]
Comment (jamesschlader) [reply backend/apis/parse-docket-pdf.utils.js]: Also noticed that too late. Will remove.

File: backend/apis/parse-docket-pdf.utils.js [hunk 7]
```diff
@@ -89,13 +89,80 @@ function parseCharges(lines, sections) {
...
+          costType = getCostType(line) || costType;
+          collection = line.includes("State Debt Collection");
```
[the remark this answers is not in the document]
Comment (jamesschlader) [reply backend/apis/parse-docket-pdf.utils.js]: Ok. I see the problem. Will fix.

File: backend/apis/parse-docket-pdf.utils.js [hunk 8]
```diff
@@ -89,13 +89,80 @@ function parseCharges(lines, sections) {
...
+        const returnObj = {
+          name: subsection.name,
+          amountDue,
+          originalAmountDue,
+          amendedAmountDue,
+          amountPaid,
+          costType,
+          collection
+        };
+        return amountDue && returnObj;
```
[the remark this answers is not in the document]
Comment (jamesschlader) [reply backend/apis/parse-docket-pdf.utils.js]: There are cases where there are too many Amount Due lines. From https://drive.google.com/drive/u/0/folders/1cW6RM4jLDeiLeSy7emI8_xdf6HhpfBhG
we have this pair of lines: 

 Original Amount Due:         600.00
 Amended Amount Due:           0.00

The code was always returning 0.00 for the Amount Due because the second line trumped the first. That's why I added the two different properties so that both pieces of information would be preserved. Such cases always have an Amount Due part on a line, so I kept that property because it was the trigger to return the returnObj on Line 151. 

If we have good reason to believe that the use of Original Amount Due and Amended Amount due are only and always used in Debt Collection cases, then we could assign the amountDue property as in the original code (returning 0.00) and make use of the collection property when testing for qualification. If not, we would need to keep both in order to make the proper calculation during the qualification step. In the latter case, I suppose the Amended Amount Due would be the most likely candidate to return as the amountDue property, but I don't know about that.

File: backend/apis/parse-docket-pdf.utils.js (hunk 1 above)
Comment (jamesschlader) [reply backend/apis/parse-docket-pdf.utils.js]: I'll look into it.

File: backend/apis/docket-pdf-test-files/get-text.js (hunk 2 above)
Comment (jamesschlader) [reply backend/apis/docket-pdf-test-files/get-text.js]: I'll delete it.

File: backend/apis/normal-docket-pdf.test.js (hunk 3 above)
Comment (jamesschlader) [reply backend/apis/normal-docket-pdf.test.js]: I'll split them up.

File: backend/apis/parse-docket-pdf.utils.js (hunk 8 above)
[the remark this answers is not in the document]
Comment (joeldenning) [reply backend/apis/parse-docket-pdf.utils.js]: 👍 yeah I saw the amendedAmountDue in the code and it seems to be a good thing to extract as a separate property.

My original question remains though - when will `amountDue` be undefined? The state debt collection cases seem to have both an `amountDue` and `amendedAmountDue`. So they don't seem like a situation where `amountDue` would be undefined.

File: backend/apis/parse-docket-pdf.utils.js (hunk 5 above)
[the remark this answers is not in the document]
Comment (joeldenning) [reply backend/apis/parse-docket-pdf.utils.js]: Ah this line in the pdf?

https://github.com/JustUtahCoders/utahexpungements.org/blob/c9522469b27675264611ce65654147627a1566bd/backend/apis/docket-pdf-test-files/account-summary.js#L46

Yeah maybe we could just ensure there's at least one alphabetical character in the line then? That way if there is a section that is only one word, the parsing would still catch it.

```
```suggestion
     /[a-zA-Z]/.test(line)
```

File: backend/apis/parse-docket-pdf.utils.js (hunk 8 above)
[the remark this answers is not in the document]
Comment (jamesschlader) [reply backend/apis/parse-docket-pdf.utils.js]: Amount Due will always be there so amountDue will never be undefined, as far as I can tell. If there's anything at all in the account summary, at least one line will have "Amount Due" on it. 

There's something that is bothering me about the code as written. We are currently getting this:
 {
      "name": "TRUST DETAIL",
      "amountDue": "0.00",
      "amountPaid": "0.00",
      "collection": false
    },

for anything that is a TRUST DETAIL. The actual raw file has this:

 TRUST DETAIL 
                 Trust Description: Interest on Rstitutn

Wouldn't we prefer to have the name = Interest on Rstitutn, or whatever comes to the right of the colon in the Trust Description line? What do you think? The eligibility calculator will need to know what the content of the TRUST DETAIL items are since the one I've printed here matters for eligibility but one like: 

 TRUST DETAIL 
                 Trust Description: Interest Attrny Fees
doesn't.

File: backend/apis/parse-docket-pdf.utils.js (hunk 1 above)
Comment (jamesschlader) [reply backend/apis/parse-docket-pdf.utils.js]: Should I add a value to costType for every case? That's a pretty straighforward task...

File: backend/apis/parse-docket-pdf.utils.js (hunk 8 above)
[the remark this answers is not in the document]
Comment (jamesschlader) [reply backend/apis/parse-docket-pdf.utils.js]:
```let trustName;
...
 if (line.includes("Trust Description")) {
            trustName = line.slice(line.indexOf(": ") + 2);
          }
...
const returnObj = {
...
trustName
}```

This code does the trick.

File: backend/apis/normal-docket-pdf.test.js (hunk 3 above)
Comment (jamesschlader) [reply backend/apis/normal-docket-pdf.test.js]: About the tests, I'm using the test example Veronica included as an example of a docket with no content in the Account Summary section. One of my test files is meant to see if the function can parse a large files traversing several pages, and the other to see if it can get the Debt Collection case right. The describe blocks are testing multiple files because they represent different edge case. 

That said, should I repeat the tests with nested describe blocks, like this:
describe("testing file A", ()=>{
   describe("testing edge case 1", ()=>{}
})

or should I just rename the files and variables so that they make more sense?

File: backend/apis/parse-docket-pdf.utils.js (hunk 1 above)
Comment (joeldenning) [reply backend/apis/parse-docket-pdf.utils.js]: We should return a cost type whenever we can derive one.

File: backend/apis/parse-docket-pdf.utils.js (hunk 8 above)
[the remark this answers is not in the document]
Comment (joeldenning) [reply backend/apis/parse-docket-pdf.utils.js]: I didn't read all of your comment because it's really long and not related to the question I asked. However, the first part of your comment is what I was asking about:

> Amount Due will always be there so amountDue will never be undefined, as far as I can tell

With that being the case, we should not have the `filter()` at all. And we should update this line of code, too, to not check amountDue for truthiness before returning it:

```suggestion
        return returnObj;
```

File: backend/apis/normal-docket-pdf.test.js (hunk 3 above)
Comment (joeldenning) [reply backend/apis/normal-docket-pdf.test.js]: You should have two `describe()` blocks: one for each test.

File: backend/apis/parse-docket-pdf.utils.js (hunk 8 above)
[the remark this answers is not in the document]
Comment (jamesschlader) [reply backend/apis/parse-docket-pdf.utils.js]: Turns out that Bail is reported without an Amount Due: 
 BAIL/CASH BOND DETAIL - TYPE: BAIL                                     
                         Posted By: PERSON
                            Posted:      1,000.00
                         Forfeited:        781.65
                          Refunded:        218.35
                           Balance:          0.00
In the absence of the amountDue check, the above pushes to the accountSummary array like this:
 {
      "name": "BAIL/CASH BONDS    Posted:      1,000.00",
      "collection": false
    },
The accountSummary array also gets the Totals sections, which lack an Amount Due, like this:
 {
      "name": "ACCOUNT SUMMARY",
      "collection": false
    },
{
      "name": "TRUST TOTALS    Trust Due:       1,050.00",
      "amountPaid": "1,050.00",
      "collection": false
    },

I must have determined that those objects were noise and therefore left them out with the line 151 in the code.

What do you think?

[the remark this answers is not in the document]
Comment (joeldenning) [reply backend/apis/parse-docket-pdf.utils.js]: I think we should not leave out any of the subsections, even if the amount due is missing

File: backend/apis/parse-docket-pdf.utils.js [hunk 9]
```diff
@@ -89,13 +89,80 @@ function parseCharges(lines, sections) {
 }
 
 function parseAccountSummary(lines, sections) {
-  const accounts = [];
+  let subsections = [];
+  let balances = [];
```
[the remark this answers is not in the document]
Comment (joeldenning) [reply backend/apis/parse-docket-pdf.utils.js]: This should still addressed

[the remark this answers is not in the document]
Comment (jamesschlader) [reply backend/apis/parse-docket-pdf.utils.js]: The balances array is the return array for the parseAccounts function. It has to be initialized outside the addAccounts function so that it can return its modified contents from parseAccounts after addAccounts has been called by the walkSection function. (That's a mouthful.)

File: backend/apis/parse-docket-pdf.utils.js (hunk 8 above)
[the remark this answers is not in the document]
Comment (jamesschlader) [reply backend/apis/parse-docket-pdf.utils.js]: Done.

File: backend/apis/parse-docket-pdf.utils.js (hunk 9 above)
[the remark this answers is not in the document]
Comment (joeldenning) [reply backend/apis/parse-docket-pdf.utils.js]: You can declare variables without initializing them. My comment here is that the `= []` is pointless, since it's always overridden on line 110 no matter what. The `[]` doesn't seem to be used in any way:

```suggestion
  let balances;
```

File: backend/apis/parse-docket-pdf.utils.js [hunk 10]
```diff
@@ -89,13 +89,102 @@ function parseCharges(lines, sections) {
...
+          amountDue,
+          originalAmountDue,
+          amendedAmountDue,
+          amountPaid,
+          costType,
+          collection
+        };
+        return returnObj;
+      })
+      .filter(result => {
```
Comment (joeldenning) [inline backend/apis/parse-docket-pdf.utils.js]: This can be removed, now that all charges are always included.

**Comment by joeldenning** [comment]
I have resolved all of the github comments that are now addressed. There are three comments remaining, and also the tests are failing. Once those comments are addressed and the build passes, I'll approve and merge.

File: backend/apis/parse-docket-pdf.utils.js (hunk 5 above)
[the remark this answers is not in the document]
Comment (jamesschlader) [reply backend/apis/parse-docket-pdf.utils.js]: I updated it to this solution.

File: backend/apis/normal-docket-pdf.test.js [hunk 11]
```diff
@@ -28,4 +34,75 @@ describe("normal docket pdf file", () => {
     expect(actual1.disposition).toBe("Dismissed (w/o prej)");
     expect(actual1.dispositionDate).toBe("Month 00, 0000");
   });
+
+  test("always adds accountSummary property as an array", () => {
+    expect(parsedText.accountSummary.length).toBe(1);
```
Comment (HappyViki) [inline backend/apis/normal-docket-pdf.test.js]: I think this should be checking if array (example: `expect(Array.isArray(parsedText.accountSummary)).toBe(true);`), instead of checking the length, because that's what the test description says.

File: backend/apis/parse-docket-pdf.utils.js [hunk 12]
```diff
@@ -89,13 +89,97 @@ function parseCharges(lines, sections) {
 }
 
 function parseAccountSummary(lines, sections) {
-  const accounts = [];
+  const subsections = [];
```
Comment (HappyViki) [inline backend/apis/parse-docket-pdf.utils.js]: You mean `let`?

Comment (HappyViki) [reply backend/apis/parse-docket-pdf.utils.js]: nvm I guess you can because it's still the same array

File: backend/apis/normal-docket-pdf.test.js (hunk 11 above)
Comment (joeldenning) [reply backend/apis/normal-docket-pdf.test.js]: I think this is fine - the `length` property is something that arrays have. Close enough 😄
